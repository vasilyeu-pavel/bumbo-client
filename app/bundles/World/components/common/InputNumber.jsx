import React, { Component, PropTypes } from 'react';
import declOfNums from '../../utils/declOfNum';
import { bindAll } from 'lodash';

export default class InputNumber extends Component {
  static propTypes = {
    titles: PropTypes.arrayOf(PropTypes.string).isRequired,
    value: PropTypes.number.isRequired,
    maxValue: PropTypes.number,
    onIncrement: PropTypes.func.isRequired,
    onDecrement: PropTypes.func.isRequired,
  }

  constructor(props, context) {
    super(props, context);

    bindAll(this, ['handleIncrement', 'handleDecrement']);
  }

  handleIncrement() {
    const { value, maxValue } = this.props;

    if (value >= maxValue) return;

    this.props.onIncrement(value + 1);
  }

  handleDecrement() {
    const { value } = this.props;

    if (value <= 1) return;

    this.props.onDecrement(value - 1);
  }

  render() {
    const { value, titles } = this.props;
    const title = declOfNums(value, titles);

    return (
      <div className='input-number'>
        <div className='input-number__decr' onClick={this.handleDecrement}>-</div>
        <div className='input-number__title'>{`${value} ${title}`}</div>
        <div className='input-number__incr' onClick={this.handleIncrement}>+</div>
      </div>
    );
  }
}
