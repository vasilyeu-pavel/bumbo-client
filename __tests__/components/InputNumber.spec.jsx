/* global expect, jest */
/* eslint arrow-body-style: 0 */

import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import InputNumber from '../../app/bundles/World/components/common/InputNumber';

describe('<InputNumber /> component', () => {
  const minProps = {
    titles: ['ночь', 'ночи', 'ночей'],
    value: 2,
    onIncrement: () => {},
    onDecrement: () => {},
  };

  it('should render component with correct title "1 ночь"', () => {
    const inputNumber = renderer.create(<InputNumber {...minProps} value={1} />).toJSON();

    expect(inputNumber).toMatchSnapshot();
  });

  it('should render component with correct title "2 ночи"', () => {
    const inputNumber = renderer.create(<InputNumber {...minProps} value={2} />).toJSON();

    expect(inputNumber).toMatchSnapshot();
  });

  it('should render component with correct title "5 ночей"', () => {
    const inputNumber = renderer.create(<InputNumber {...minProps} value={5} />).toJSON();

    expect(inputNumber).toMatchSnapshot();
  });

  it('should increment value after click on increment button', () => {
    const handleChange = (value) => expect(value).toEqual(minProps.value + 1);
    const handler = jest.fn(handleChange);
    const inputNumber = shallow(<InputNumber {...minProps} onIncrement={handler} />);
    const incrementButton = inputNumber.find('.input-number__incr');

    incrementButton.simulate('click');

    expect(handler).toHaveBeenCalled();
  });

  it('should not call "onIncrement" method if value >= maxValue', () => {
    const props = { value: 10, maxValue: 10 };
    const handler = jest.fn();
    const inputNumber = shallow(<InputNumber {...minProps} {...props} onIncrement={handler} />);
    const incrementButton = inputNumber.find('.input-number__incr');

    incrementButton.simulate('click');

    expect(handler).not.toHaveBeenCalled();
  });

  it('should decrement value after click on decrement button', () => {
    const handleChange = (value) => expect(value).toEqual(minProps.value - 1);
    const handler = jest.fn(handleChange);
    const inputNumber = shallow(<InputNumber {...minProps} onDecrement={handler} />);
    const decrementButton = inputNumber.find('.input-number__decr');

    decrementButton.simulate('click');

    expect(handler).toHaveBeenCalled();
  });

  it('should not call "onDecrement" method if value <= 1', () => {
    const handler = jest.fn();
    const inputNumber = shallow(<InputNumber {...minProps} value={1} onDecrement={handler} />);
    const decrementButton = inputNumber.find('.input-number__decr');

    decrementButton.simulate('click');

    expect(handler).not.toHaveBeenCalled();
  });
});
