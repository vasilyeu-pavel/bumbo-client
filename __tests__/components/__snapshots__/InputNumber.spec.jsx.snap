exports[`<InputNumber /> component should render component with correct title "1 ночь" 1`] = `
<div
  className="input-number">
  <div
    className="input-number__decr"
    onClick={[Function]}>
    -
  </div>
  <div
    className="input-number__title">
    1 ночь
  </div>
  <div
    className="input-number__incr"
    onClick={[Function]}>
    +
  </div>
</div>
`;

exports[`<InputNumber /> component should render component with correct title "2 ночи" 1`] = `
<div
  className="input-number">
  <div
    className="input-number__decr"
    onClick={[Function]}>
    -
  </div>
  <div
    className="input-number__title">
    2 ночи
  </div>
  <div
    className="input-number__incr"
    onClick={[Function]}>
    +
  </div>
</div>
`;

exports[`<InputNumber /> component should render component with correct title "5 ночей" 1`] = `
<div
  className="input-number">
  <div
    className="input-number__decr"
    onClick={[Function]}>
    -
  </div>
  <div
    className="input-number__title">
    5 ночей
  </div>
  <div
    className="input-number__incr"
    onClick={[Function]}>
    +
  </div>
</div>
`;
